analysis/refactor/rewrite/rewriters.go

```diff
@@ -0,0 +1,32 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rewrite
+
+import (
+	"github.com/awslabs/ar-go-tools/internal/rewrite"
+	"golang.org/x/tools/go/packages"
+)
+
+// ApplyRewrites applies a series of rewrite steps to the packages and all its dependencies.
+// The rewrites are meant to preserve soundness of the analysis, under the condition that the focus points of the
+// analysis are not the entrypoint or endpoints of analyses.
+// The rewrites are:
+//
+// All the rewrites from the base rewrites used in capslock package:
+// - calls to [sort.Slice] and [sort.SliceStable] are removed and replaced by calls to the functions used in sorting.
+// - calls to (sync.Once).Do are replaced by a call to the method in the sync.Once object.
```

Comment:
We have to a bit careful here because this rewrite removes the writes performed by e.g. `Once.Do` to internal data. If we are relying on the pointer analysis to track all writes and there is a once.do in some protected space, this rewrite would miss it. In theory incorrect pointer analysis results can propagate arbitrarily far but I'm not sure I could construct an example of actual unsoundness from these functions, but I haven't considered it very carefully.

A more obviously sound version would be to clone the methods in question, but that would be significantly more complicated to implement.

Reply:
I think in general that is true, but I don't think we can construct such an example with `sync.Once` for dataflow analysis, unless those objects are specifically the ones being tracked, or `Do` is marked as a sink/source. A `sync.Once` has some private fields which are not aliased in the implementation and the client can't alias them.
I will probably add more of those cases and also reason about them.